spyder_kernels/comms/commbase.py

```diff
@@ -0,0 +1,344 @@
+        - The 'content' is a dict with: {
+                'is_error': a boolean indicating if the return value is an
+                            exception to be raised.
+                'call_id': The uuid from above,
+                'call_name': The function name (mostly for debugging)
+                }
+
+"""
+import cloudpickle
+import logging
+import sys
+import uuid
+import time
+
+PY2 = sys.version[0] == '2'
+logger = logging.getLogger(__name__)
+# To be able to get and set variables between Python 2 and 3
+PICKLE_PROTOCOL = 2
+TIMEOUT = 10
+
+
+class CommBase(object):
+    """
+    Class with the necessary attributes and methods to handle
+    communications with a console.
+    The subclass must open a comm somehow and register it with
+    self._register_comm
+    """
+
+    def __init__(self):
+        self._comm = None
+        self._message_handlers = {}
+        self._remote_call_handlers = {}
+        self._call_reply_dict = {}
+        self.register_message_handler(
+            'remote_call', self._handle_remote_call)
+        self.register_message_handler(
+            'blocking_call_reply', self._handle_blocking_call_reply)
+
+    @property
+    def comm_name(self):
+        return 'spyder_api'
+
+    def close(self):
+        """Close the comm."""
+        self._comm.close()
+        self._comm = None
+
+    def is_open(self):
+        """Check to see it the comm is open."""
+        return self._comm is not None
+
+    def register_message_handler(self, message_id, handler):
+        """
+        Register a message handler.
+
+        Parameters
+        ----------
+        message_id : str
+            The identifier for the message
+        handler : callback
+            A function to handle the message. This is called with 3 arguments:
+                - msg_dict # A dictionnary with messages informations
+                - buffer # The data transmitted in the buffer
+                - load_exception # exception from buffer unpickling
+            Pass None to unregister the message_id
+        """
+        if not handler:
+            self._message_handlers.pop(message_id, None)
+            return
+
+        self._message_handlers[message_id] = handler
+
+    def register_call_handler(self, call_name, handler):
+        """
+        Register a remote call handler.
+
+        This is a more specialised version of the message handler.
+        This acts similarly to a function call.
+
+        Parameters
+        ----------
+        call_name : str
+            The name of the called function
+        handler : callback
+            A function to handle the request.
+            Pass None to unregister the call_name
+        """
+        if not handler:
+            self._remote_call_handlers.pop(call_name, None)
+            return
+
+        self._remote_call_handlers[call_name] = handler
+
+    def send_message(self, spyder_msg_type, content=None, data=None):
```

Comment:
The `send_message` is almost a duplicate of the `remote_call` mechanism. As "There should be only one obvious way to do it", should `send_message` and `register_message_handler` be made private?

Reply:
I will do that as well